### PR TITLE
add avoid_js_rounded_ints

### LINKED ISSUE #955: dart2js lint when integer literals will have rounding errors
Add a lint to indicate that​ certain integer constants are not precisely represented when compiled to JS. For example, 1000000000000000001 is a 59-bit odd number, so is rounded to 1000000000000000000 fit in the effective-53-bits of the double.

### PR DESCRIPTION
**Pull request by a14n**
Follow up on #957.

Fixes #955

### CODE REVIEW

**Comment by a14n** [comment]
cc @rakudrama

File: lib/src/rules/avoid_js_rounded_ints.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Avoid JavaScript rounded ints.';
+
+const _details = r'''
+
+**AVOID** integers that will be rounded.
```
Comment (bwilkerson) [inline lib/src/rules/avoid_js_rounded_ints.dart]: Perhaps "integers" --> "integer literals"

Comment (a14n) [reply lib/src/rules/avoid_js_rounded_ints.dart]: done

Comment (rakudrama) [reply lib/src/rules/avoid_js_rounded_ints.dart]: This is more precise: **AVOID** integer literals that cannot be represented exactly when compiled to JavaScript.

File: lib/src/rules/avoid_js_rounded_ints.dart [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+When a program is compiled to JavaScript `int` and `double` become JavaScript
+Numbers. Too large integers (`value < Number.MIN_SAFE_INTEGER` or
+`value > Number.MAX_SAFE_INTEGER`) may be rounded to the closest Number value.
```
Comment (rakudrama) [inline lib/src/rules/avoid_js_rounded_ints.dart]: Is it possible to put actual values in the message? It would make the message more concrete.
If you could say something like the following, that would make the situation clearer:

1000000000000000001 cannot be represented exactly as a JavaScript Number, so 1000000000000000000 will be used instead.

Comment (rakudrama) [reply lib/src/rules/avoid_js_rounded_ints.dart]: Note that the nearest representable number is `value.toDouble()`, but this might be a value that cannot be represented as an `int` (when 0x7ff....fff is rounded up to 0x800....000), but you could print it with
`new BigInt.from(value.toDouble()).toString()`

File: test/rules/avoid_js_rounded_ints.dart [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+final i = 1; // OK
+final min = -9007199254740991; // OK
+final max = 9007199254740991; // OK
+
+final minErr = -9007199254740993; // LINT
+final maxErr = 9007199254740993; // LINT
+
+final notRounded = 1000000000000000000; // OK
+final rounded = 1000000000000000001; // LINT
```
Comment (rakudrama) [inline test/rules/avoid_js_rounded_ints.dart]: Also test with -9223372036854775808
value.abs() for this number is negative on the 64-bit integer VM.
Lucky it is not rounded! (-2^63), so maybe comment that the code works for this edge case.

File: lib/src/rules/avoid_js_rounded_ints.dart (hunk 1 above)
Comment (a14n) [reply lib/src/rules/avoid_js_rounded_ints.dart]: done

File: lib/src/rules/avoid_js_rounded_ints.dart (hunk 2 above)
Comment (a14n) [reply lib/src/rules/avoid_js_rounded_ints.dart]: >Is it possible to put actual values in the message?

It's not something that linter currently does. It would need some changes if we want to do that. 
@pq @bwilkerson what do you think about customizing messages depending on the code linted ?

File: test/rules/avoid_js_rounded_ints.dart (hunk 3 above)
Comment (a14n) [reply test/rules/avoid_js_rounded_ints.dart]: done

@bwilkerson `9223372036854775808` has `node.value == null`. Is it a bug ?

File: lib/src/rules/avoid_js_rounded_ints.dart (hunk 2 above)
Comment (bwilkerson) [reply lib/src/rules/avoid_js_rounded_ints.dart]: Actually, I think it's possible to do that today, it's just that nobody has done it so there's no pattern to follow. I'd be fine with merging this change and then refining it, or you can make this change before merging if you'd like.

The trick is that you are not restricted to using `reportLint` to report a lint; I believe you have access to the error reporter that `reportLint` uses so you can create your own LintCode with a message that can be customized by using the substitution mechanism supported by errors. For example, the message might look something like `"The value '{0}' will be rounded to {1} when compiled to JavaScript."`.

**Comment by bwilkerson** [comment]
@rakudrama Does this look good to you?

**Comment by rakudrama** [comment]
Looks good to me.  I would still like to see the user's values used in the message to make the problem concrete (for example, they could replace the original number with the rounded one in mock data, rather than guessing what the rounded result is). I'm ok with that as a followup change.

File: test/rules/avoid_js_rounded_ints.dart (hunk 3 above)
Comment (bwilkerson) [reply test/rules/avoid_js_rounded_ints.dart]: It sounds like a bug. Every integer literal that is a valid 64-bit integer ought to have a non-null value.

**Comment by a14n** [comment]
>Did you want to update the error message before we merge this?

I can but the bug I mentioned about `node.value == null` will cause issue.

**Comment by bwilkerson** [comment]
Good point. I'll merge it now and we can update the message after we've fixed that bug. Is it line 15 that illustrates the problem?

**Comment by a14n** [comment]
problem occurs on line 22

>final absNegative = -9223372036854775808;
